sources/vega/snapshot/v1/snapshot.proto

```diff
@@ -553,9 +553,25 @@ message StakeVerifierPending {
   string id = 8;
 }
 
-message Topology{
+message KeyRotation {
+  uint64 block_height = 1;
+  string node_id = 2;
+  string new_pub_key = 3;
+  string old_pub_key = 4;
+}
+
+message PendingKeyRotation {
+  uint64 block_height = 1;
+  string node_id = 2;
+  string new_pub_key = 3;
+  uint32 new_key_number = 4;
+}
+
+message Topology {
   repeated vega.events.v1.ValidatorUpdate validator_data = 1;
   repeated string chain_keys = 2;
+  repeated PendingKeyRotation pending_pub_key_rotations = 3;
+  repeated KeyRotation processed_pub_key_rotations = 4;
```

Comment:
What is the reason for core keeping all the key rotations that have been processed?

Reply:
So people can query historically. Because from a certain point the validator pub key will change and all rewards and delegations has to be changed to. It would be good to get users an overview of how did that happen.

Reply:
Is that not something that should be sent and stored in the datanode?

Reply:
Well currently we are in some weird kind of hybrid case where users can query some crucial data on core (I think this is quite important). It is also being send to data node btw.

But I am not super fussed about it.

https://github.com/vegaprotocol/protos/blob/develop/sources/vega/api/v1/corestate.proto#L12

Reply:
Actually looking again here: https://github.com/vegaprotocol/vega/blob/develop/coreapi/service.go
These endpoints just subscribe to the broker and collect the information from events, they don't read historical state stored in core.

Reply:
Yeah fair enough.. I have removed it. Thanks for checking